versioning/worker-versions.md

```diff
@@ -0,0 +1,137 @@
+
+It also requires updating clients who are starting workflows to point at the
+newer task queue name.
+
+**Versioned workflow names**
+This approach is much like versioning task queue names, but trades off some of
+the operational complexity for some code maintenance complexity.
+
+Rather than needing to keep old workers around, users now need to keep the
+old code around until no more workflows using the old type are open.
+
+They still have the same problem with needing to tell clients to use the new
+workflow name.
+
+## Improving the situation
+
+The `patch` type APIs can't be eliminated completely. There will always be 
+situations where a "broken" workflow needs to be fixed to make progress. Such
+fixes are typically less common than just making changes to workflow logic, though.
+
+To make the more common case easier, we will provide a set of features that
+make the "versioned task queues" approach much easier to use for changes to
+workflows which break history compatability (but not interface compatability).
+
+At a high level:
+* Provide a simple API for versioning entire workers (this proposal)
+* Allow clients to point at task queue `foo`, and automatically route new workflow executions
+  to the latest version of the `foo` queue.
+* Provide good notifications and information about when old workers may be
+  decommissioned
+
+Building on this base set of features, we plan to eventually implement
+automatic safe-rollout of changes, and possibly tools like k8s operators
+which can automatically manage worker deployments.
+
+## SDK API changes
+
+The meat of this proposal is simple:
+**Add a version field to worker options in each SDK**
+
+Because version information will need to be sent to the Temporal Cluster,
+we define it as a protobuf message (which will show up in our API somewhere):
+```protobuf
+message WorkerVersions {
+    UserVersion worker_version = 1;
+}
+
+message UserVersion {
+    // The string the user passed as their version representation
+    string original = 1;
+    // Parsed from original string
+    uint32 major = 2;
+    // Parsed from original string
+    uint32 minor = 3;
+}
+```
+
+This message is pre-emptively named plurally in order to accommodate potential
+increases in versioning granularity in the future. For example, we have considered
+versioning Workflows, Activities, and Data Converters/Interceptors all separately.
+They could be added as new fields to this message if/when that happens.
+
+Each language will expose a way to (optionally) set the worker version to 
+start, supporting the possibility of other versions later. This will translate 
+into the above message to be sent to server.
+
+### TypeScript
+
+```typescript
+Worker.create({ workerVersion: '1.0', ...otherOptions })
+```
+
+### Go
+
+```go
+workerOptions := worker.Options{
+  WorkerVersion: "1.0",
+  // other options
+}
+```
+
+### Java
+
+```java
+WorkerOptions.newBuilder().setWorkerVersion("1.0").build()
```

Comment:
Do we ever consider the possibility that the author of the worker startup code is distinct from the author of the workflow? I could see an infra team being in charge of the "workflow platform" and an appdev team implementing the workflows themselves. IIUC this proposal has a tacit assumption that the workflow author is always simultaneously in charge of worker startup - is that correct?

Reply:
Yeah, that situation argues for the "version every workflow function individually" approach. Honestly in many ways I prefer that but we decided to go with this for it's simplicity and the fact that it looks consistent in every language / is easier to introduce in a nonbreaking way.

The unfortunate result of that is yes you've gotta know about underlying changes in that situation.

Reply:
Also worth nothing - we could always layer on that approach later if it becomes clear it's desired. You could start using per-wf-type versions and that could get aggregated up into a `workflowsVersions` mapping which is used in a similar way to this version number.

Reply:
Yeah, I would have preferred the individual approach but I get your reasoning as well.